Switch `DataModel::Provider` interface to default enabled (except on linux where it stays as check)

andy31415 opened the pull request:
This will switch from using ember-compatibility-functions.cpp to using the codegen datamodel provider.

PR should show FLASH and RAM sideffect as well as if any compilation issues may occur.

#### Changes

- enable data model interface usage by default (except linux where the state remains "check" and runs both in parallel)
- Fix build issues and bugs (some split in other PRs). Specifically I have found:
   - k32w0 is very close to full flash, so enabled progress-level logs and above only (no detail)
   - fixed bug in caching being maintained across dynamic endpoint enable/disable (if the same endpoint id is re-used, the same path may reference freed clusters)

### Size notes

- The PR has a non-trivial size increase, generally due to making iteration over attributes & commands configurable rather than fixed-index iteration. 
- It started off at around 1% of extra flash, after which several PRs were created to address sizes and reduce this, including:
  - #36249
  - #36247
  - #36246
  - #36229
  - #36172 
  - #36273
- The unmerged above PRs are included in this change to evaluate the current size overhead.
- Additional size reductions (globally) in PRs:
  - #36253
  - #36251 offers some minute size decreases (however in the data_model implementation those savings seem to zero out)
  - #36286
  - #36288

andreilitvin committed:
Switch data model provider usage to default to enabled

andreilitvin committed:
Decouple pump configuration and control from compatibility functions. Still 100% ember tied.

andy31415 commented:
chip-k32w0x-light-example does not fit in flash anymore ... that is not good at all.

Looking through the rest of the failures and will have separate PRs to fix them.

andy31415 commented:
Converted to draft: we need to figure out all build failures and even then figure out how to manage flash increases.

andreilitvin committed:
Fix up some private/public dependencies: instance header should always be a private dependency

andreilitvin committed:
Allow java (or generally dynamic server) controller builds to compile with dynamic dispatcher. Dynamic dispatcher seems to only care about commands so implemented all the rest as empty, which is not ideal

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by gn

andreilitvin committed:
Better dynamic dispatch implementation, removed extra TODO functions

restyled-commits committed:
Restyled by whitespace

andreilitvin committed:
Update dep a bit more

andreilitvin committed:
Fix linter

andreilitvin committed:
Fix linter again

andreilitvin committed:
Update NXP to support various logging targets and for K32W0 enable progress logging and up (no detail/automation)

andreilitvin committed:
Fix nxp compile (ordering issue)

andreilitvin committed:
Update logic a bit to use log level for low power

andreilitvin committed:
Update builds to consider low power already removing logs

andy31415 committed:
Merge branch 'master' into default_enable_dm_provider

andy31415 committed:
Make microwave oven cluster compilable

andy31415 committed:
Restyle

andy31415 committed:
Add codegen data model to darwin xproject: given that we add ember, we need these to access the ember bits as there is no app or data model to provide these

andy31415 committed:
Add missing override for darwin dispatch when codegen data model uses lower level methods for ember access

andy31415 committed:
Fix linter

andy31415 committed:
Add ability to track structural changes in ember for the purpose of codegen data model caching

andy31415 committed:
Fix typo: uint is unsigned

andy31415 committed:
Restyle

andy31415 committed:
Fix android to also provide a generation version

andy31415 committed:
Fix another uint/unsigned typo ... oops

bzbarsky-apple commented on src/darwin/Framework/CHIP/ServerEndpoint/MTRIMDispatch.mm:
```diff
@@ -50,6 +50,13 @@ void emberAfClusterInitCallback(EndpointId endpoint, ClusterId clusterId)
 
+Protocols::InteractionModel::Status emberAfWriteAttribute(const ConcreteAttributePath & path, const EmberAfWriteDataInput & input)
+{
+    // All of our attributes are handled via AttributeAccessInterface, so this
```
Should have assertChipStackLockedByCurrentThread(), no?

andy31415 committed:
Add assertion of stack lock in darwin override of emberAfWriteAttribute

andy31415 committed:
Merge branch 'master' into default_enable_dm_provider

andy31415 commented:
#36104 contains the split out change for `metadata structure generation`

andy31415 committed:
Merge branch 'master' into default_enable_dm_provider

restyled-commits committed:
Restyled by isort

andy31415 committed:
Remove formatting changes from python files

andy31415 committed:
Pick up restyle changes too

andy31415 committed:
Merge branch 'master' into default_enable_dm_provider

andy31415 committed:
Merge branch 'master' into default_enable_dm_provider

andy31415 committed:
Use TLV-based encoding for ember data buffer.

This saves 2K of flash on some test devices, being much more
efficient in code size compared to using datamodel code.

tehampson commented on .github/workflows/examples-nxp.yaml:
```diff
@@ -70,7 +70,7 @@ jobs:
                   scripts/run_in_build_env.sh "\
                       ./scripts/build/build_examples.py \
-                      --target nxp-k32w0-freertos-lighting-factory \
+                      --target nxp-k32w0-freertos-lighting-factory-log-progress \
```
Is this supposed to be in this PR?

andy31415 replied:
Yes, k32w0 is low on flash and this makes CI fit by disabling detail log strings. Without this, this PR flash increases do not pass CI.

andy31415 replied:
Added a comment explaining.

andy31415 commented on src/app/dynamic_server/mock_includes/zap-generated/endpoint_config.h:
```diff
@@ -21,3 +21,5 @@
 #define GENERATED_DEFAULTS #error "Not a real codegen. This is a temporary include for dynamic-overrides only"
 #define GENERATED_CLUSTERS #error "Not a real codegen. This is a temporary include for dynamic-overrides only"
+
+#define ATTRIBUTE_LARGEST (1003)
```
src/app/util/ember-io-storage.cpp allocates size for attributes based on this. I will add a comment.

andy31415 committed:
Add comment on attribute laregest

andy31415 committed:
Merge branch 'master' into default_enable_dm_provider

bzbarsky-apple commented on .github/workflows/examples-nxp.yaml:
```diff
@@ -70,7 +70,7 @@ jobs:
                   scripts/run_in_build_env.sh "\
                       ./scripts/build/build_examples.py \
-                      --target nxp-k32w0-freertos-lighting-factory \
+                      --target nxp-k32w0-freertos-lighting-factory-log-progress \
```
What is the current flash increase from turning this on?  What steps are we taking to bring it down?

Ideally, we would bring it down _before_ enabling it on constrained platforms.

andy31415 replied:
started with #36172

In the end the extra code is to support interfaces that are not `index in a contiguous array allocated in ram`.

andy31415 replied:
#36229 is a followup, saves another 1K of flash.

andy31415 replied:
Resolving as I started a list of flash savings PRs in the summary of this PR.

bzbarsky-apple commented on src/app/BUILD.gn:
```diff
@@ -225,14 +225,17 @@ static_library("interaction-model") {
 
   deps = [ "${chip_root}/src/app:events" ]
 
+  # Temporary dependency: codegen data provider instance should be provided
+  # by the application
```
Where is the issue tracking this?  It should be linked here.

andy31415 replied:
#36230 is the generic coverage for these updates.

bzbarsky-apple commented on src/app/BUILD.gn:
```diff
@@ -294,13 +297,23 @@ static_library("interaction-model") {
+    # DynamicDispatcher is actually an ember-override that takes over
+    # util/attribute-storage.cpp and util/attribute-table.cpp functions.
+    #
+    # We likely should formalize and change this with a proper DataModel::Provider that
```
DynamicDispatcher should be switched to use a DataModel::Provider, indeed.  The problem is right now it has no way to provide one, right?

Need tracking issues for:

1. A way to inject a data model provider during stack bringup instead of defaulting in the codegen one.  For DynamicDispatcher, having this just on the controller factory would be enough.
2. Making DynamicDispatcher actually use a provider.

Both should be linked from here, presumably.

andy31415 replied:
I cannot do this in one PR: existing code assumes embed is built in and everything evolves around that. My plan would be that every app would have to set a provider before continuing and the default provider (if any) should just be a noop/deny everything.

Before doing that though we have to change every application to set a provider and for controller sides we have to figure out where such provider is set (probably in some sort of platform init call).

andy31415 replied:
Created #36230 for a generic coverage of things. I will split it up when starting to work on it - for now just keeping the list of checkmarks of todo items.

bzbarsky-apple commented on src/app/clusters/microwave-oven-control-server/microwave-oven-control-server.cpp:
```diff
@@ -248,11 +248,16 @@ void Instance::HandleSetCookingParameters(HandlerContext & ctx, const Commands::
     {
+        ConcreteCommandPath commandPath(mEndpointId, OperationalState::Id, OperationalState::Commands::Start::Id);
+
+#if CHIP_CONFIG_USE_EMBER_DATA_MODEL
```
This seems like we should just have some sort of "server command exists" API on the IM engine, no?

bzbarsky-apple commented on src/app/dynamic_server/DynamicDispatcher.cpp:
```diff
@@ -381,3 +385,62 @@ const EmberAfCluster * emberAfFindServerCluster(EndpointId endpoint, ClusterId c
+
+unsigned emberAfMetadataStructureGeneration()
+{
+    // DynamicServer at this point hardcodes a single OTA provider cluster
```
What is DynamicServer?

bzbarsky-apple commented on src/darwin/Framework/Matter.xcodeproj/project.pbxproj:
Yeah, all this stuff desperately needs to go away and we need to have a way to inject a provider.  But for now, I guess it will have to do.

bzbarsky-apple reviewed (changes requested):
I meant to request changes.  I don't think the flash increases here are acceptable, some of the avenues we discussed for reducing those involve changing the provider APIs, and I don't think we want to enable and start using this while we are still figuring the API out.

restyled-commits committed:
Restyled by clang-format

restyled-commits committed:
Restyled by shfmt

andy31415 committed:
Undo unrelated change

andy31415 committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

andy31415 committed:
Add some casts to make android compiler happy

andy31415 committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

andy31415 committed:
Update darwin build project as well with the new files

andy31415 committed:
Updates based on code review

andy31415 committed:
Added unit tests for min/max int64 values

andy31415 committed:
Rename PascalString to PascalStringType

andy31415 committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

andy31415 committed:
Fix rename

andy31415 committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

andy31415 committed:
Restyle

andy31415 committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

andreilitvin committed:
Add helper methods inside odd sized integers to localize code logic

andreilitvin committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

restyled-commits committed:
Restyled by clang-format

andreilitvin committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

andreilitvin committed:
Fix up negative ranges

andreilitvin committed:
Fixed ranges

andreilitvin committed:
Fix signed max bug and update unit tests

andreilitvin committed:
Merge branch 'master' into ember_tlv_reader_encoding

andreilitvin committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

andreilitvin committed:
Make android happy

andreilitvin committed:
Typo fix

andreilitvin committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

andy31415 committed:
Switch up unit tests

andy31415 committed:
Merge branch 'master' into ember_tlv_reader_encoding

andy31415 committed:
Update a nullable check to make use of ValueToNullValue

andy31415 committed:
Add namespace prefix

andy31415 committed:
Update src/app/codegen-data-model-provider/EmberDataBuffer.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

andy31415 committed:
Update src/app/codegen-data-model-provider/EmberDataBuffer.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

andreilitvin committed:
Correct comments:signed, not unsigned

andreilitvin committed:
Use constructors for the buffer info

andreilitvin committed:
Rename things to EmberAttributeDataBuffer

andreilitvin committed:
Merge branch 'master' into ember_tlv_reader_encoding

andreilitvin committed:
Undo submodule updates

andreilitvin committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

restyled-commits committed:
Restyled by clang-format

andreilitvin committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

andreilitvin committed:
Merge branch 'master' into ember_tlv_reader_encoding

andreilitvin committed:
Merge branch 'ember_tlv_reader_encoding' into default_enable_dm_provider

andreilitvin committed:
Use EmberAttributeDataBuffer for codegen provider _Read

andreilitvin committed:
Merge branch 'master' into ember_tlv_writer_encoding

andreilitvin committed:
Fix comments

restyled-commits committed:
Restyled by clang-format

andreilitvin committed:
Merge branch 'ember_tlv_writer_encoding' into default_enable_dm_provider

andreilitvin committed:
Merge branch 'ember_tlv_writer_encoding' into default_enable_dm_provider

andreilitvin committed:
Update DynamicDispatcher comment

andreilitvin committed:
Rename file

andreilitvin committed:
Merge branch 'master' into ember_tlv_writer_encoding

andreilitvin committed:
Merge branch 'ember_tlv_writer_encoding' into default_enable_dm_provider

andreilitvin committed:
Merge branch 'master' into default_enable_dm_provider

andreilitvin committed:
Use a pointer for the subject descriptor.

This seems to save about 88 bytes of flash on a test NRF board.

andreilitvin committed:
Restyle

andreilitvin committed:
Fix include

andreilitvin committed:
Fix include

andreilitvin committed:
Merge branch 'use_pointer_for_subject_descriptor' into default_enable_dm_provider

andreilitvin committed:
This saves 136 bytes...

andreilitvin committed:
More changes to save slightly more flash for code

andreilitvin committed:
Restyle

andreilitvin committed:
Fix typo

andreilitvin committed:
Also fix PW rpc

andreilitvin committed:
Fix includes

andreilitvin committed:
Merge branch 'test_first_next_updates' into default_enable_dm_provider

andreilitvin committed:
Merge branch 'use_pointer_for_subject_descriptor' into default_enable_dm_provider

andreilitvin committed:
Merge branch 'master' into default_enable_dm_provider

andreilitvin committed:
make more detailed logging optional in the codegen data model and enable it only on known large platforms

andreilitvin committed:
Merge branch 'remove_logging_from_codegen_dm' into default_enable_dm_provider

andreilitvin committed:
Pull back the event path validity mixin, start with a datamodel implementation

andreilitvin committed:
Fix compile logic after I moved things away

andreilitvin committed:
Add one more check

andreilitvin committed:
Restyle

andreilitvin committed:
Fix typo

andreilitvin committed:
Fix includes

andreilitvin committed:
Restyle

andreilitvin committed:
Move decodable lists bits as a non-template class to save flash

andy31415 committed:
Update src/app/InteractionModelEngine.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

andy31415 committed:
Rename method

andy31415 committed:
Restyle

andy31415 committed:
More renames

andy31415 committed:
Restyle

andy31415 committed:
Fix some renames

andy31415 committed:
Restyle

andy31415 committed:
Merge branch 'master' into untemplate_parts_of_decodable_list

andy31415 committed:
Merge branch 'untemplate_parts_of_decodable_list' into default_enable_dm_provider

andy31415 committed:
A few more renames

andy31415 committed:
Restyle

andy31415 committed:
Merge branch 'cleanup_event_logic' into default_enable_dm_provider

andy31415 committed:
Use chip::app::IsSignedAttributeType

andy31415 committed:
Fix up put as well as naming for null value and comment

andy31415 committed:
Fix up nullable tests

andy31415 committed:
Test that you cannot decode a null value for non-nullable double and single

andy31415 committed:
Allow NAN for non-nullable floating points

andy31415 committed:
Add test case for non nullable bool

andy31415 committed:
Restyle

andy31415 committed:
Merge branch 'post_merge_cleanup_36229_ember_buffer_encode' into default_enable_dm_provider

andy31415 committed:
Add a header for efr32

andy31415 committed:
Update src/app/codegen-data-model-provider/EmberAttributeDataBuffer.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

andy31415 committed:
Update src/app/codegen-data-model-provider/EmberAttributeDataBuffer.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

andy31415 committed:
Remove extra comment

andy31415 committed:
Replace switch with if

andy31415 committed:
Comment fix

andy31415 committed:
Merge branch 'post_merge_cleanup_36229_ember_buffer_encode' into default_enable_dm_provider

andy31415 committed:
Merge branch 'master' into default_enable_dm_provider

andy31415 committed:
Merge branch 'master' into post_merge_cleanup_36229_ember_buffer_encode

andy31415 committed:
Another try to make efr32 build of tests happy

andy31415 committed:
Move includes around, to try to work around issues within efr32 compiles...

andy31415 committed:
more updates, this time local efr32 compiles

andy31415 committed:
Merge branch 'post_merge_cleanup_36229_ember_buffer_encode' into default_enable_dm_provider

andy31415 committed:
Remove lookup tables from ember attribute data buffer

andy31415 committed:
Merge branch 'even_smaller_ember_code' into default_enable_dm_provider

andy31415 committed:
Merge branch 'master' into default_enable_dm_provider

andreilitvin committed:
Merge branch 'master' into default_enable_dm_provider
